pkg/sql/alter_index.go

```diff
@@ -0,0 +1,146 @@
+	indexDesc, err = tableDesc.FindIndexByID(indexDesc.ID)
+	if err != nil {
+		return nil, err
+	}
+	return &alterIndexNode{n: n, tableDesc: tableDesc, indexDesc: indexDesc}, nil
+}
+
+func (n *alterIndexNode) startExec(params runParams) error {
+	// Commands can either change the descriptor directly (for
+	// alterations that don't require a backfill) or add a mutation to
+	// the list.
+	descriptorChanged := false
+	origNumMutations := len(n.tableDesc.Mutations)
+
+	for _, cmd := range n.n.Cmds {
+		switch t := cmd.(type) {
+		case *tree.AlterIndexPartitionBy:
+			previousTableDesc := *n.tableDesc
+			partitioning, _, err := createPartitionedBy(
+				params.ctx, params.p.evalCtx.Settings, &params.p.evalCtx, n.tableDesc, n.indexDesc, t.PartitionBy)
+			if err != nil {
+				return err
+			}
+			n.indexDesc.Partitioning = partitioning
+			// TODO(benesch): install CHECK constraint to exclude rows that do
+			// not belong to any partition of the unique constraint.
+			//
+			// TODO(dan): This checks TableDescriptors instead of
+			// PartitioningDescriptors to mirror the fast path check. Of course,
+			// RepartitioningFastPathAvailable could also be acting on
+			// PartitioningDescriptors but then it would need a complicated
+			// method signature and it felt weird to impose that on the
+			// sql<->sqlccl hook. Revisit?
+			descriptorChanged = !proto.Equal(&previousTableDesc, n.tableDesc)
+			if descriptorChanged {
+				// TODO(dan): Consider the case of a table that is repartitioned
+				// when it has not finished the schema change for adding the
+				// partitioning CHECK constraint from the last partitioning.
+				fastPath, err := RepartitioningFastPathAvailable(&previousTableDesc, n.tableDesc)
+				if err != nil {
+					return err
+				}
+				if !fastPath {
+					return fmt.Errorf("data validation is required for this repartitioning, but is currently unimplemented")
+				}
+				// TODO(dan): Remove zone configs which no longer point at a
+				// partition. This also needs to be done for indexes.
+			}
+		default:
+			return fmt.Errorf("unsupported alter cmd: %T", cmd)
+		}
+	}
+
+	if err := n.tableDesc.AllocateIDs(); err != nil {
+		return err
+	}
+
+	addedMutations := len(n.tableDesc.Mutations) > origNumMutations
+	mutationID := sqlbase.InvalidMutationID
+	var err error
+	if addedMutations {
+		mutationID, err = params.p.createSchemaChangeJob(params.ctx, n.tableDesc,
+			tree.AsStringWithFlags(n.n, tree.FmtSimpleQualified))
+	} else {
+		err = n.tableDesc.SetUpVersion()
+	}
+	if err != nil {
+		return err
+	}
+
+	if err := params.p.writeTableDesc(params.ctx, n.tableDesc); err != nil {
+		return err
+	}
+
+	// Record this index alteration in the event log. This is an auditable log
+	// event and is recorded in the same transaction as the table descriptor
+	// update.
+	if err := MakeEventLogger(params.p.LeaseMgr()).InsertEventRecord(
+		params.ctx,
+		params.p.txn,
+		// TODO(dan): BEFORE MERGE make an EventLogAlterIndex to use here
```

Comment:
Intentionally leaving this TODO undone until I make sure that no-one is going to have strong objections to adding this generic CreateIndex